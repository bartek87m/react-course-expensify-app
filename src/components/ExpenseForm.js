import React from 'react';
import moment from 'moment';
import 'react-dates/initialize';
import { SingleDatePicker } from 'react-dates';


// const date = new Date();
const now = moment();
console.log(now.format("MMM Do, YYYY"));

export default class ExpenseForm extends React.Component{
   constructor(props){//pobranie propsów
    super(props); //przekazujemy je do super
    console.log(props);
    this.state = {
        description: props.expense ? props.expense.description: '',
        note: props.expense ? props.expense.note: '',
        amount: props.expense ? (props.expense.amount / 100).toString(): '',
        createdAd: props.expense ? moment(props.expense.createdAd): moment(),
        calendarFocused: false,
        error: ''
    }

   }
 
  
    onDescriptionChange = (e) => {
        const description = e.target.value;
        this.setState(() => ({description}))
    }

    onNoteChange = (e) => {
        const note = e.target.value;
        this.setState(() => ({note}))
    }

    onAmountChange = (e) => {
        const amount = e.target.value;
        if(!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)){//wyrażenie rególarne
             this.setState(() => ({amount}));
        }
    }

    onDataChange = (createdAd) => {
        if(createdAd) {
            this.setState(() => ({createdAd}))
        }
        
    }

    onFocusChange = ({focused}) => {
        this.setState(() => ({calendarFocused: focused}))
    }

    onSubmit = (e) => {
        e.preventDefault();

        if (!this.state.description || !this.state.amount) {
            this.setState(() => ({error: 'Please enter valid data'}));
        }else{
            this.setState(() => ({error: ''}));
            this.props.onSubmit({
                description: this.state.description,
                amount: parseFloat(this.state.amount, 10) * 100,
                createdAd: this.state.createdAd.valueOf(),
                note: this.state.note


            });
        }
    }

    render(){
        return(
            <div>
                {this.state.error && <p>{this.state.error}</p>}
                <form onSubmit={this.onSubmit}>
                    <input
                        type="text"
                        placeholder="Description"
                        autoFocus //przy wejściu na stronę automatycznie ustawia kursor na ten input  
                        value={this.state.description}
                        onChange = {this.onDescriptionChange}
                        />
                    <input
                        type="number"
                        placeholder="Amount"
                        value={this.state.amount}
                        onChange={this.onAmountChange}
                    />
                    <SingleDatePicker
                        date = {this.state.createdAd}
                        onDateChange =  {this.onDataChange}
                        focused = {this.state.calendarFocused}
                        onFocusChange = {this.onFocusChange}
                        numberOfMonths={1}
                        isOutsideRange={() => false}
                    />

                    <textarea
                        placeholder="Add a note for your expense (optional)"
                        value={this.state.note}
                        onChange={this.onNoteChange}
                        />
                    <button>Add Expense</button>
                    
                    </form>
            </div>
        )
    }
}