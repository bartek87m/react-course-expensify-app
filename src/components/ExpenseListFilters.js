
import React from 'react';
import {connect} from 'react-redux';
import 'react-dates/initialize';
// import 'react-dates/lib/css/_datepicker.css';
import { DateRangePicker } from 'react-dates';
import { setTextFilter, sortByDate, sortByAmount, setStartDate, setEndDate } from '../actions/filters';


export class ExpenseListFilter extends React.Component{ //exportowanie nie połaczonej z reduxem wersji

    state = {
        focusedInput: null
    };

    onDatesChange = ({startDate, endDate}) => {
         this.props.setStartDate(startDate);
         this.props.setEndDate(endDate);

    };

    onTextChange = (e) => {
        this.props.setTextFilter(e.target.value);
    };

    onSortChange = (e) => {
        if(e.target.value === 'amount'){
            this.props.sortByAmount();
        }else if(e.target.value === 'date'){
            this.props.sortByDate();
        };
    };

    render(){
        return(
            <div className="content-container">
                <div className="input-group">
                    <div className="input-group__item">
                        <input 
                            className="text-input" 
                            placeholder="Search expense"
                            type="text" 
                            value={this.props.filters.text} 
                            onChange={this.onTextChange}/>
                    </div>
                    <div className="input-group__item">
                        <select className="select" value={this.props.filters.sortBy} onChange={this.onSortChange}>
                            <option value='date'>Date</option>
                            <option value='amount'>Amount</option>
                        </select>
                    </div>
                    <div className="input-group__item">
                        <DateRangePicker
                            startDate={this.props.filters.startDate}
                            startDateId="startDateId"
                            endDate={this.props.filters.endDate}
                            endDateId="endDateId"  
                            onDatesChange={this.onDatesChange} 
                            focusedInput={this.state.focusedInput}
                            onFocusChange={focusedInput => this.setState({ focusedInput })} 
                            showClearDates={true}
                            numberOfMonths={1}
                            isOutsideRange = {() => false}
                        />
                    </div>
                </div>            
            </div>
        )
    }
}

const mapStateToProps = (state) => ({filters: state.filters});

const mapDispatchToProps = (dispatch) => ({
    setTextFilter: (text) => dispatch(setTextFilter(text)),
    sortByDate: () => dispatch(sortByDate()),
    sortByAmount: () => dispatch(sortByAmount()),
    setStartDate: (startDate) => dispatch(setStartDate(startDate)),
    setEndDate: (endDate) => dispatch(setEndDate(endDate))

});

export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilter); 
//wyeksportowanie wersji połączonej z redux