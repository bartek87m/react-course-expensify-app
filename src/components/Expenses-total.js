import React from 'react';
import expenses from "../tests/fixtures/expenses";



class Expenses_totat extends React.Component {

    calculateCosts = (expenses) => {
   
        if(expenses.length === 0){
            return 0;
        }else if(expenses.length === 1){
            return expenses[0].amount;
        }else if(expenses.length > 1){
            return expenses.map((expense) =>  expense.amount).reduce((accumulator, currentValue) => accumulator + currentValue);
        }
    }  

    render() {

        return(
            <p>{this.calculateCosts(expenses)}</p>
        ) 
    }
   
}

export default Expenses_totat;