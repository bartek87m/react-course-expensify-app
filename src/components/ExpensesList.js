import React from 'react';
import {connect} from 'react-redux';
import ExpenseListItem from './ExpenseListItem'
import selecrtedExpenses from '../selectors/expenses'

export const ExpenseList = (props) => {
                return (
    <div>
        {
            props.expenses.length === 0 ?(
                <p>No Expenses</p>
            ) : (
                props.expenses.map((expense) => {
                   return <ExpenseListItem key={expense.id} {...expense}/>   //teraz ExpenseListItem ma dostemp do całego expense
                })
    
       

        )}
        
    </div>

);
            };

//połaczeni komponentu ExpenseList do redux, ExpenseList ma teraz dostę do store
// i przez props może używac danych ze store
//kiedy store się zmienia  komponent zmienia się dynamicznie tez dynamicznie się zmienia
const mapStateProps = (state) => {
    return {
        expenses: selecrtedExpenses(state.expenses, state.filters)
    };
}

export default connect(mapStateProps)(ExpenseList); //łączymy komponent z żeczami, które chcemy dostać ze store

