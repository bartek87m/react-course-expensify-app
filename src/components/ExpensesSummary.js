import React from 'react'
import { connect } from 'react-redux';
import total from "../selectors/total";
import selecrtedExpenses from '../selectors/expenses'
import numeral from 'numeral';

export const ExpensesSummary = (props) => {
       
   
    console.log("form Summary", props.visibleExpenses2);
        
        let expenseCount = props.visibleExpenses.length;
        let expensesTotal = numeral(props.expensesTotal / 100).format("$0,0.00");

        return (
            <div>
                <p>Viewing {expenseCount} expenses totalling {expensesTotal}</p>
            </div>
            )
    }
    

const mapStateProps = (state) => {
    const visibleExpenses = selecrtedExpenses(state.expenses, state.filters)
    return {
        visibleExpenses: selecrtedExpenses(state.expenses, state.filters),
        expensesTotal: total(visibleExpenses)
    };
}


export default connect(mapStateProps)(ExpensesSummary); //higher component