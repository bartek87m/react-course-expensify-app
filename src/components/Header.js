import React from 'react';
import {NavLink} from 'react-router-dom';
import { connect } from 'react-redux';
import { startLogout } from '../actions/auth';

export const Header = (props) => (
    <header>
        <h1>Expensifity</h1>
        <NavLink to="/" exact={true} activeClassName="is-active">Home Page</NavLink>
        <NavLink to="/create" exact={true} activeClassName="is-active">Create expense</NavLink>
        <NavLink to="/help" exact={true} activeClassName="is-active">Help Page</NavLink>
        <button onClick={props.startLogout}>Logout</button>
    </header>
);

const mapDipatchToProps = (dispatch) => ({ //dostÄ™p do dispatch
    startLogout: () => dispatch(startLogout())
})

export default connect(undefined, mapDipatchToProps)(Header);