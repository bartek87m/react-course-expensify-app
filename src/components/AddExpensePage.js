import React from 'react';
import {connect} from 'react-redux';
import ExpenseForm from "./ExpenseForm";
import {startAddExpense} from '../actions/expenses';

export class AddExpensePage extends React.Component{
    onSubmit = (expense) => {
        // props.dispatch(addExpense(expense));
        console.log(expense)
        this.props.startAddExpense(expense)
        this.props.history.push('/');//tak jakby ktoś kliknoł link
    } 
    render() {
        return(
            <div>
            <h1>Add Expense</h1>
            <ExpenseForm
                onSubmit={this.onSubmit}
            />
        </div> 
        )
    }
}

const mapDispatchToProps = (dispatch) => ({ //pozwala zwrócić dispathc poza komponentem
    startAddExpense: (expense) => dispatch(startAddExpense(expense))
    
})

export default connect(undefined, mapDispatchToProps)(AddExpensePage);