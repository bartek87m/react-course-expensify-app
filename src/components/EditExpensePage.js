import React from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm'
import {editExpense, removeExpense} from '../actions/expenses';

export class EditExpensePage extends React.Component{

    onSubmit = (expense) => {//to się dzieje kiedy ExpenseForm jest submitowana:)
        // props.dispatch(editExpense(props.expense.id, expense))
        this.props.editExpense(this.props.expense.id, expense);
        this.props.history.push('/')//przekierowanie
    }

    onClick = () => {
        // props.dispatch(removeExpense({ id: props.expense.id}));
        // console.log("ID: " + this.props.expense.id);
        this.props.removeExpense(this.props.expense.id);
        this.props.history.push('/')//przekierowanie
    }

    render() {
        return(

            <div>
                <ExpenseForm
                    expense = {this.props.expense}//przekazanie danych uzyskanych z redux
                    onSubmit= {this.onSubmit}
                />
                <button onClick={this.onClick}>Remove</button>
            </div>
        )
        
    }
}

const mapStateToProps = (state, props) => ({
    
        expense:state.expenses.find((expense) => expense.id === props.match.params.id)
             //propsy z higher component
        
    
});

const mapDispatchToProps = (dispatch, props) => ({
    
        editExpense: (id, expense) => dispatch(editExpense(id, expense)),
        removeExpense: (data) => dispatch(removeExpense({id: data}))
        
}); 

//dzięki connect poprzez props mamy dostęp do dispatch
export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage); //higher component