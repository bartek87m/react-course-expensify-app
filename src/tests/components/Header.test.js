import React from 'react';
import { shallow } from 'enzyme';
import toJSON from 'enzyme-to-json';
import { Header } from '../../components/Header';
//'enzyme-to-json' pozwala na uzyskanie czystego snapshota bez zbędnych danych
//react-test-render

//shallow tylko renderuje podany komponent
//full dom render komponuje równiech komponenty child

test('should render Header correctly', () => {
    const wrapper = shallow(<Header startLogout={() => {}}/>);
    expect(wrapper).toMatchSnapshot();

});

test('should simulate Header ', () => {
    const startLogout = jest.fn();
    const wrapper = shallow(<Header startLogout={startLogout}/>);
    wrapper.find('button').simulate('click');
    expect(startLogout).toHaveBeenCalled();
});