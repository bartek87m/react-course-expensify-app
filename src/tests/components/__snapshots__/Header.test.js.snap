// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render Header correctly 1`] = `
<header>
  <h1>
    Expensifity
  </h1>
  <NavLink
    activeClassName="is-active"
    aria-current="page"
    exact={true}
    to="/"
  >
    Home Page
  </NavLink>
  <NavLink
    activeClassName="is-active"
    aria-current="page"
    exact={true}
    to="/create"
  >
    Create expense
  </NavLink>
  <NavLink
    activeClassName="is-active"
    aria-current="page"
    exact={true}
    to="/help"
  >
    Help Page
  </NavLink>
  <button
    onClick={[Function]}
  >
    Logout
  </button>
</header>
`;
